### DISCUSSION TITLE
How to replicate equivalent to PS cmdlets Get-DnsServerZone and Get-DnsServerResourceRecord

### DISCUSSION THREAD

**Discussion by flipcode1973** [MichaCo/DnsClient.NET#188] [General]
Hi, great library. Been trying to replicate an existing PS script in C# that retrieves DNS zones and resource records e.g.
```

$zones = Get-DnsServerZone -ComputerName $DNSServer
$allColumns = @()

$zones | ForEach-Object {
  if ($_.ZoneType -eq 'Forwarder' -or $_.ZoneType -eq 'Primary'){
    $zone = $_.ZoneName
    Write-Output "`n[$(Get-Date)] - Checking $zone for any records"

    $records_from_zone = Get-DnsServerResourceRecord -ComputerName $DNSServer -ZoneName $zone # | Where-Object { $_.RecordType -eq  'A' -or $_.RecordType -eq 'CNAME' }
    Write-Output "[$(Get-Date)] - $($records_from_zone.Count) records found! preparing to process, please wait..."
```

  but am struggling to see how to do this using DnsClient.NET.  Am I just being thick? (every possibility) :-)

**Comment by MichaConrad** [comment]
maybe check this out: https://github.com/rmbolger/DnsClient-PS

I personally don't do much with powershel, so I cannot really help you ;)
Getting zone information is not really that easy though outside of DNS Server adminstrative tools where you have full access. You can request a *zone transfer* with a dnsclient (like this library) but you might not get any results unless you have access to the server.

**Reply by flipcode1973** [reply]
Thanks for the quick response Michael, I'll have a look at the link you sent through. 

Perhaps my question wasn't clear, I don't want to use your library in PS, I want to replace the existing PS routine with one in C# (utilising your library). 

When you say 'request a zone transfer', what function/option would achieve this? 

And similarly, once I retrieve the zones (if available), how would I then get the properties (e.g Hostname, DistinguishedName, RecordType etc) like that of **Get-DnsServerResourceRecord**? Is this even achievable outside the PS cmdlets? Cheers Colin

**Comment by MichaConrad** [comment]
Ah ok, no that was not clear. Then I don't really get what the problem is. The documentation should have enough examples of how to query records.

Just query for A type records or whatever you need using whatever server you want to use.

Regarding zone information, the DNS protocol has a query type `AXFR` which is supported by this library if TCP transport is used.

**Reply by flipcode1973** [reply]
Thanks for the info, I'll dig into the docs. Must be doing something wrong as I'm getting a 'Non-Existent Domain' error when I try querying with QueryType.AXFR.

**Comment by MichaConrad** [comment]
Well, as I said, TCP mode must be enabled for AXFR, that's just required per RFC -> `DnsQuerySettings.UseTcpOnly`
And, the you need permissions to get a reply from the DNS server you are asking. Otherwise, you'll not get anything back.

**Reply by flipcode1973** [reply]
Yeah, I suspect its permissions. Thanks for all your help.
